crates/cext/src/circuit.rs

```diff
@@ -0,0 +1,264 @@
+    gate: StandardGate,
+    qubits: *const u32,
+    params: *const f64,
+) -> ExitCode {
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
+    unsafe {
+        let qargs: &[Qubit] = match gate.num_qubits() {
+            0 => &[],
+            1 => &[Qubit(*qubits.wrapping_add(0))],
+            2 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+            ],
+            3 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+                Qubit(*qubits.wrapping_add(2)),
+            ],
+            // There are no standard gates > 3 qubits
+            _ => unreachable!(),
+        };
+        let params: &[Param] = match gate.num_params() {
+            0 => &[],
+            1 => &[(*params.wrapping_add(0)).into()],
+            2 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+            ],
+            3 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+                (*params.wrapping_add(2)).into(),
+            ],
+            // There are no standard gates that take > 3 params
+            _ => unreachable!(),
+        };
+        circuit.push_standard_gate(gate, params, qargs);
+    }
+    ExitCode::Success
+}
+
+#[repr(C)]
+pub struct OpCount {
+    name: *mut c_char,
+    count: usize,
+}
+
+#[repr(C)]
+pub struct OpCounts {
+    data: *mut OpCount,
+    len: usize,
+}
+
+/// @ingroup QkCircuit
+/// Return a list of string names for instructions in a circuit and their counts
+///
+/// @param circuit A pointer to the circuit to get the counts for
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_count_ops(qc);
+///
+/// # Safety
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> OpCounts {
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+    let count_ops = circuit.count_ops();
+    let mut output: Vec<OpCount> = count_ops
+        .into_iter()
+        .map(|(name, count)| OpCount {
+            name: CString::new(name).unwrap().into_raw(),
+            count,
+        })
+        .collect();
+    let data = output.as_mut_ptr();
+    let len = output.len();
+    std::mem::forget(output);
+    OpCounts { data, len }
+}
+
+/// @ingroup QkCircuit
+/// Free a circuit op count list.
+///
+/// @param op_counts The returned op count list from ``qk_circuit_count_ops``
+///
+/// # Safety
+///
+/// Behavior is undefined if ``op_counts`` is not the object returned by ``qk_circuit_count_ops``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_free_count_ops(op_counts: OpCounts) {
```

Comment:
Maybe 
```suggestion
pub unsafe extern "C" fn qk_opcounts_free(op_counts: OpCounts) {
```
since the scheme is `qk_<object>_free`?

Reply:
Renamed in: https://github.com/Qiskit/qiskit/pull/14006/commits/3e62ec6d8d4b68bb1521de0660659c214b7afa83